core/src/main/scala/org/apache/spark/sql/execution/ObjectHashMapAccessor.scala

```diff
@@ -304,32 +305,42 @@ case class ObjectHashMapAccessor(@transient session: SnappySession,
     val updateMinMax = integralKeys.map { index =>
       s"$hashMapTerm.updateLimits(${keyVars(index).value}, $index);"
     }.mkString("\n")
+    // for SnappyData column or row tables there is no need to make
+    // a copy of non-primitive values into the map
+    var hasExchange = false
+    val doCopy = child.find {
```

Comment:
I am quite sure that's not the case. The find above is a short-cut for:
```scala
child.find { p match {
  case ...
} }
```
which matches the first pattern and returns true/false. What this find will do is return the first PartitionedPhysicalPlan. If no such is found then "doCopy" should be true as noted in the comments (that strings are immutable for our tables but may not be for any others) hence the isEmpty.

The bug was in the exchange check. I have moved out the variable as an additional condition on top in last push -- please see.

However, I do agree this check is a fragile one and can lead to problems for more complicated joins that do not have an exchange. Let me think more about it how to make it tighter and deterministic.